rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,370 @@
+
+- the original discovered resource name: `.Name`
+- cloud metadata: `.AWS`, `.GCP`, `.Azure`
+  - each of these should only be available for the corresponding matcher
+    type: `aws:`, `gcp:`, `azure:`.
+- the region/location of the resource: `.Region` or `.Location`
+  - this is already included in AWS metadata, but not Azure nor GCP.
+  - rather than a protobuf update, we can just expose this template data
+    variable separately: `.Region` for AWS/Azure and `.Location` for GCP, for
+    consistency with the naming conventions chosen for our discovery matcher
+    config - AWS/Azure matchers use `regions:` and GCP matchers use
+    `locations:`.
+
+For each cloud metadata type, we can make available the corresponding
+`api/types` protobufs:
+
+- `api/types.AWS`
+- `api/types.GCPCloudSQL`
+- `api/types.Azure`
+
+Doing this would couple our cloud metadata protobufs to the supported template
+variables, however it would simplify the implementation and ensure that any new
+cloud metadata is available for templating.
+
+For whatever choice of supported variables, we must document the supported
+template variables in our docs reference.
+
+We should also print a friendly user message that lists the supported
+variables if a user provides a template string that references an unsupported
+variable.
+
+We can check the user provided template when file config is applied,
+or when a dynamic discovery config is created, by executing the template with
+stub data input, or by walking the template parse tree.
+
+### Security
+
+This configuration setting will only be available to Teleport cluster admins,
+which limits the potential for intentional abuse. I'm not aware of any
+unintentional security concerns, since we control the available supported
+template variables and none of them are secrets.
+
+I was concerned about the potential for resource exhaustion if a
+self-referencing template makes it into a running discovery agent:
+`{{template "ResourceNameTemplate" .}}` - this is self-referencing and looks
+like it will evaluate infinitely.
+What will actually happen is text/template terminates the recursion at 100,000
+depth with an error (not a panic).
+
+We could alternatively catch the invocation of `{{template}} "name" .` actions 
+by writing a visitor to walk the template parse tree. We could use this parse
+tree walking for further template restrictions if we need to.
+
+We can check for invocations of the template action on agent startup from static
+config, or server-side when a dynamic discovery config is created.
+
+### UX
+
+#### `discovery_service` new service configuration properties
+
+Users need to ensure the `resource_name_template` option is set to a template
+string for each discovery matcher to enable resource name rewriting.
+
+Example:
+
+```yaml
+discovery_service:
+  enabled: true
+  aws:
+    - types: ["ec2", "rds"]
+      regions: ["us-west-1"]
+      resource_name_template: "{{.Name}}-{{.Region}}-{{.AWS.AccountID}}"
```

Comment:
question - why do we need to give users such a flexibility. Why can't we always import resource names with name, region and account ID in mind as the default, making sure the names are unique? 

Do we really need such flexibility and complexity? To me that looks like opportunity for all sorts of issues @GavinFrazar @r0mant

Reply:
We could do that, there were a couple of reasons we wanted to go with the templating approach, mostly UX and compatibility related:

- If we always import with name/region/accountID, imported database names will become very long/unwieldy and users might have a hard time using them in e.g. `tsh db connect`.
- By default we'll still default to just "name" like now so only users who have the name conflict problem will need to use this.

IIRC we discussed this on the product sync a while ago but we can reconsider if you still think this is a bad idea.

Reply:
I remember this conversation. But now seeing the proposal, I understand how needlessly complicated the whole setup would be.

```
Name                             Description                      Allowed Users       Labels          Connect                          
-------------------------------- -------------------------------- ------------------- --------------- -------------------------------- 
> sales-center-prod-reader-ek... Salescenter AWS RDS PostgreSQ... [platform_readonly] app=sales-ce... tsh db connect sales-center-p... 
sales-center-staging-reader-e... Salescenter AWS RDS PostgreSQ... [platform_readonly] app=sales-ce...             
```

This is a pretty long name, but it does not create any issues for me. 

I wonder if there is easier way to solve this problem using just minor UX tweaks.

For example, what if we set name to the unique name at all times, but also show a short name, derived from the original resource name and allow connecting by labels as well:

```
Name                                                                     Full name                          
--------------------------------                        ------------------- 
mydatabase                                                           mydatabase-us-east-1-1234-567-8910
```  

Then resource name will be long name nobody will read, and if there is no collision, then `tsh db connect mydatabase` will work exactly as `tsh db connect mydatabase-us-east-1-1234-567-8910` If there is a collision we will ask users to provide fully qualified name.

We can also let users `tsh db connect label=value` as well.

What do you think?

Reply:
@klizhentas  I like the idea to default to a long name and print a short name instead of templating, since teleport config is already very complex.

I think supporting `tsh db connect label=value` would be useful, in and of itself, for executing statements on multiple databases at once non-interactively, which is something I wanted before for setting up Azure MySQL IAM users in multiple databases at once:

```
bash-5.1$ cat <<'EOF' | tsh db connect --db-user=database-access 'engine="Microsoft.DBforMySQL/servers"'
SET aad_auth_validate_oids_in_tenant = OFF;
CREATE AADUSER 'SomeNewUser' IDENTIFIED BY '64ca60b1-5b77-4dc1-9ad3-79ac7fb1fd09'; # this is the ID of the service principal
GRANT ALL ON `%`.* TO 'SomeNewUser'@'%';
EOF           
```

Reply:
I like the idea of allowing `tsh db connect` to connect using substring of a name and labels 👍 This should also be backwards compatible since "instance name" will be a part of the whole name as well so users will still be able to connect using it unless there's ambiguity.

Then it sounds like we don't need any config changes, just these 3 UX tweaks:

1. When importing databases, include account ID and region in the resource name.
2. Update `tsh db connect` and `tsh proxy db` to treat provided "name" as a substring of the resource name. Return error in case of ambiguity.
3. Update `tsh db connect` and `tsh proxy db` to select database by labels. Return error in case of ambiguity.

For "2" and "3", we return error in case of ambiguity asking user to specify "unambiguated" (is that a word lol?) name, and maybe including a list of names that matched in the message.

Does that sound reasonable? @GavinFrazar @klizhentas @smallinsky

Reply:
sounds good to me, essentially what `tsh ssh` does already - shall I close this RFD then? or change it to what you just proposed?

Reply:
sounds good to me, but @r0mant you'd have to include UI and connect in scope of this change as well so we don't break things.

Reply:
If we apply it for databases, we should also apply for Kubernetes discovery. Currently, it supports all the same features that database discovery supports. it also allows replacing the cluster name with an import rule.

If in the future we extend it to other clouds, we need to take into account that Azure subscription ids are UUIDs and it's pretty difficult to remember which UUID corresponds to dev/prod/stg accounts. AWS account ids are shorter and simpler but UUIDs are messy and difficult to remember.

Besides, we need to decide if we need to keep the compatibility of cached certificates. For example, tsh keeps Kubernetes certs cached locally and the certificate contains the Kubernetes cluster name. If we going to continue with this change and since discovered clusters are dynamically loaded, users might end up with cached certificates (for a long time) that cannot be used for anything because the embedded cluster name does not match with the server's clusters names.

Reply:
@r0mant @klizhentas 
I did some investigation into using label selectors, and I think my proposal in the comments above to support executing commands on multiple matching databases at once will require a lot more consideration to actually implement.

Therefore, to limit the scope of this RFD to just what is necessary to avoid name collisions in discovered resources, I've rewritten the RFD to just support prefixes and label matching to unambiguously select a single resource for now.

@tigrato 
> Besides, we need to decide if we need to keep the compatibility of cached certificates. For example, tsh keeps Kubernetes certs cached locally and the certificate contains the Kubernetes cluster name. If we going to continue with this change and since discovered clusters are dynamically loaded, users might end up with cached certificates (for a long time) that cannot be used for anything because the embedded cluster name does not match with the server's clusters names.

Could you expand on that? I don't understand how the scenario you've described applies - the cert will have the full kube cluster name in it, we'll just support short names for identifying kube clusters when using `tsh kube login`

Reply:
My concern is that:

1. A user has an auto-discovery config in place and working
2. calls `tsh kube login <cluster_name>`
3. Accesses the cluster and caches the certificate locally `kubernetes_cluster=<cluster_name>`
4. Someone upgrades `discovery_service` to a version that implements this RFD and the cluster names change. What was `cluster_name` now becomes `{account_id}-{region}-{cluster_name}`.
5. The next call to kubectl will fail because the cluster embedded in the certificate - `<cluster_name>` - no longer exists.  To fix it, it requires changes to the kubeconfig or calling `tsh kube login` again

Since `tsh kube credentials` does not validate if the cluster exists when the certificate is cached to reduce the execution time, the error returned will be not found from the `kubernetes proxy` which is sub-optimal.

Reply:
>  proposal in the comments above to support executing commands on multiple matching databases at once will require a lot more consideration to actually implement.

@GavinFrazar I might have missed it, but why do we need to execute commands on multiple databases? People have asked about something like this before but this is out of scope of this issue. Let's skip it for now.

@tigrato I think it's ok to reserve this change for Teleport 14 and not complicate implementation with trying to preserve compatibility for certificates. People will have to relogin, yes, but that should be fine and we can mention it in the T14 release notes. We've done things that require relogin before. WDYT?

Reply:
> To clarify though: it's not just prefixes, it's any substring of the name, correct?

I think prefix matching will be more intuitive than substrings in `tsh`. Consider this example if we do substrings:
```shell
# foo-rds-us-west-1-0123456789012 # this one made by discovery
# rds-prod-something # this one made by static/dynamic config

$ tsh db connect rds
error: ambiguous ...*snip*...
```

But I think substring search in Connect/WebUI makes sense because the user first searches and then interactively selects the resource to use

Reply:
> @tigrato I think it's ok to reserve this change for Teleport 14 and not complicate implementation with trying to preserve compatibility for certificates. People will have to relogin, yes, but that should be fine and we can mention it in the T14 release notes. We've done things that require relogin before. WDYT?

I am ok with this as long as we are clear about the change and include it in the breaking changes notes. 

Handling the compatibility of certificates would be a nightmare for security reasons and we shouldn't pursue that path.